axiom/nr_segment_types.h

```diff
@@ -0,0 +1,207 @@
+  nrtime_t total_time;
+  nr_exclusive_time_t* main_context;
+} nr_segment_tree_to_heap_metadata_t;
+
+/*
+ * Segment Coloring
+ *
+ * The C Agent API gives customers the ability to arbitrarily parent a segment
+ * with any other segment. As a result, it is possible to introduce a cycle
+ * into the tree. To avoid infinite regress during the recursive traversal of
+ * the tree, the nodes are colored during traversal to indicate that they've
+ * already been traversed.
+ */
+typedef enum _nr_segment_color_t {
+  NR_SEGMENT_WHITE,
+  NR_SEGMENT_GREY
+} nr_segment_color_t;
+
+/*
+ * Segment priority indicators
+ *
+ * These go into the priority bitfield in the nr_segment_t struct and can be
+ * set via nr_segment_set_priority_flag. The higher the value of the priority
+ * field, the higher the likelihood that the span created from the segment will
+ * be kept.
+ *
+ * NR_SEGMENT_PRIORITY_ROOT indicates that the segment in question is a root
+ * segment.
+ *
+ * NR_SEGMENT_PRIORITY_DT indicates that the segment's id is included in an
+ * outbound DT payload.
+ *
+ * NR_SEGMENT_PRIORITY_LOG indicates that the segment's id is included in an
+ * log payload.
+ *
+ * NR_SEGMENT_PRIORITY_ATTR indicates that the segment has user attributes
+ * added to it.
+ */
+#define NR_SEGMENT_PRIORITY_ROOT (1 << 16)
+#define NR_SEGMENT_PRIORITY_DT (1 << 15)
+#define NR_SEGMENT_PRIORITY_LOG (1 << 14)
+#define NR_SEGMENT_PRIORITY_ATTR (1 << 13)
+
+typedef struct _nr_segment_datastore_t {
+  char* component; /* The name of the database vendor or driver */
+  char* sql;
+  char* sql_obfuscated;
+  char* input_query_json;
+  char* backtrace_json;
+  char* explain_plan_json;
+  nr_datastore_instance_t instance;
+} nr_segment_datastore_t;
+
+typedef struct _nr_segment_external_t {
+  char* transaction_guid;
+  char* uri;
+  char* library;
+  char* procedure; /* Also known as method. */
+  uint64_t status;
+} nr_segment_external_t;
+
+typedef struct _nr_segment_metric_t {
+  char* name;
+  bool scoped;
+} nr_segment_metric_t;
+
+typedef struct _nr_segment_error_t {
+  char* error_message; /* The error message that will appear on a span event. */
+  char* error_class;   /* The error class that will appear on a span event. */
+} nr_segment_error_t;
+
+/*
+ * Type specific fields.
+ *
+ * The union type can only hold one struct at a time. This ensures that we
+ * will not reserve memory for variables that are not applicable for this type
+ * of node. Example: A datastore node will not need to store a method and an
+ * external node will not need to store a component.
+ *
+ * You must check the nr_segment_type to determine which struct is being used.
+ */
+typedef union {
+  nr_segment_datastore_t datastore;
+  nr_segment_external_t external;
+} nr_segment_typed_attributes_t;
+
+typedef struct _nr_segment_t {
+  nr_segment_type_t type;
+  nrtxn_t* txn;
+
+  /* Tree related stuff. */
+  nr_segment_t* parent;
+  nr_segment_children_t children;
+  size_t child_ix;
```

Comment:
Introduction of this new field can be easily missed by reviewer amidst the code refactor. `child_ix` is an index into segment's parent children array (either segment's `parent->children.vector` or segment's `parent->children.packed.elements`).

Reply:
removed the refactor in favor of forward declared getters/setters in [3b715b4](https://github.com/newrelic/newrelic-php-agent/pull/801/commits/3b715b452a87bd96c81f5d7e51589bb2d173f67a)